### PR TITLE
Moved ModalFog component as a generic component

### PR DESCRIPTION
**Pull request by wakooka**
## Description

Moved ModalFog component as a generic component so we can use it anywhere.

Kapture 2021-03-18 at 17 16 46

## What has been done

- Created `Fog`
- Unit tests to make sure we can't click on elements underneath `Fog` when it is in loading state 
 
## Things to consider

- Added a polyfill on `inert` attr for a11y since it's still an experimental feature in all browsers

## How it was tested

Storybook + Unit test

### CODE REVIEW

**Comment by nickoola** [comment]
Really cool new component! 
I'll checkout your branch to see if the `inert` attribute is correctly interpreted by screen reader (if it skip content as expected).
For me it's a layout behavior and should be part of Pixar Layout specs. Is it the case in your confluence doc?

File: src/fog/Fog.unit.tsx [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+  it('should not be able to click on an interactive element', () => {
+    const onClick = jest.fn()
+    render(
+      <Fog isLoading>
+        <button onClick={onClick}>Button</button>
+      </Fog>,
+    )
+
+    const button = screen.getByRole('button')
+    expect(onClick).not.toHaveBeenCalled()
```
Comment (wakooka) [inline src/fog/Fog.unit.tsx:30]: Oops I forgot to `fireEvent.click(button)`  that's why the is passing 🤦‍♂️

**Comment by wakooka** [comment]
> Really cool new component!
> I'll checkout your branch to see if the `inert` attribute is correctly interpreted by screen reader (if it skip content as expected).

It should be ok because I'm also adding a `aria-hidden` to true

> For me it's a layout behavior and should be part of Pixar Layout specs. Is it the case in your confluence doc?

For now it's way too early to think about this component as part of the layout specs. The layout specs today are there for having sticky & position logic basically, but it  might evolve in the future in something more detailed.

File: src/fog/Fog.unit.tsx (hunk 1 above)
Comment (wakooka) [reply src/fog/Fog.unit.tsx]: I updated the test in my last commit, since I'm using `aria-hidden`, the button shouldn't even be "readable"

**Comment by nickoola** [comment]
> > Really cool new component!
> > I'll checkout your branch to see if the `inert` attribute is correctly interpreted by screen reader (if it skip content as expected).
> 
> It should be ok because I'm also adding a `aria-hidden` to true
> 
> > For me it's a layout behavior and should be part of Pixar Layout specs. Is it the case in your confluence doc?
> 
> For now it's way too early to think about this component as part of the layout specs. The layout specs today are there for having sticky & position logic basically, but it might evolve in the future in something more detailed.

Everything works fine with Screen reader.

Just have a question on this extra markup, I remark there is an empty `<div aria-hidden="true">` before the `inert` wrapper:
Screenshot 2021-03-19 at 11 01 54
Could is be removed?

**Comment by wakooka** [comment]
>  Could it be removed?

@nickoola  no, this element is the one on top with the semi-opacity (see gif)

**Review by nickoola — APPROVED** [review]

File: src/fog/Fog.tsx [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+  isLoading?: boolean
+  children: React.ReactNode
+}>
+
+export const Fog = ({ isLoading, children }: FogProps) => (
+  <StyledFogContainer>
+    <Fragment>
+      <StyledFog $isLoading={isLoading} aria-hidden="true" />
+      <div
+        ref={node => node && isLoading && node.setAttribute('inert', '')}
```
Comment (simonrelet) [inline src/fog/Fog.tsx]: Don't we want to remove the `inert` attribute if `isLoading` is `false`?

File: src/fog/Fog.tsx [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+
+export type FogProps = Readonly<{
+  isLoading?: boolean
+  children: React.ReactNode
+}>
+
+export const Fog = ({ isLoading, children }: FogProps) => (
+  <StyledFogContainer>
+    <StyledFog $isLoading={isLoading} aria-hidden="true" />
+    <div ref={node => node && isLoading && node.setAttribute('inert', '')} aria-hidden={isLoading}>
```
Comment (simonrelet) [inline src/fog/Fog.tsx:15]: Don't we want to remove the `inert` attribute when `isLoading` is set back to `false`?

**Review by simonrelet — APPROVED** [review]

File: src/fog/Fog.tsx (hunk 3 above)
Comment (wakooka) [reply src/fog/Fog.tsx:15]: By looking at the devTools it does, but I'm wondering if it's the way storybook renders the component. Maybe I should add a else with an explicite `removeAttribute`

Comment (simonrelet) [reply src/fog/Fog.tsx:15]: Maybe it's because React set all the attributes during render (remove old ones) and then the ref callback is called.

Comment (wakooka) [reply src/fog/Fog.tsx]: Done 👍

**Pull request merged**